web/html/src/manager/minion/packages/package-states.js

```diff
@@ -1,399 +1,363 @@
-          return state;
-        })
+const action = {
+  SAVE: "Save",
+  APPLY: "Apply",
+  GETSERVERPACKAGES: "GetServerPackages",
+  SEARCH: "Search"
+};
+
+const PackageStates = ({serverId}) => {
+  const [filter, setFilter] = useState<string>("");
+  const [view, setView] = useState<string>("system");
+  const [tableRows, setTableRows] = useState<Array<PackagesObject>>([]);
+  const [changed, setChanged] = useImmer<ChangesMapObject>({});
+  const searchRef = useRef<AsyncButton>();
+
+  const {
+    messages, fetchPackageStatesApi, packageStates, searchResults
+  } = usePackageStatesApi();
+
+  useEffect(() => {
+    fetchPackageStatesApi(action.GETSERVERPACKAGES, serverId)
+      .catch((error => {
+        showErrorToastr(error, {autoHide: false});
+      }));
+  }, []);
+
+  useEffect(() => {
+    generateTableData();
+  }, [changed, packageStates, searchResults, view]);
+
+  function addChanged(original: Package, newPackageStateId: OptionalValue, newVersionConstraintId: OptionalValue): void {
+    const key = packageHelpers.packageStateKey(original);
+    const currentState = changed[key];
+    if (currentState !== undefined
+      && newPackageStateId === currentState.original.packageStateId
+      && newVersionConstraintId === currentState.original.versionConstraintId) {
+      setChanged((draft: ChangesMapObject) => {
+        delete draft[key];
       });
-    });
-  }
-
-  triggerSearch() {
-    this.searchButton.trigger()
-  }
-
-  search() {
-    if (this.state.filter === this.state.search.filter) {
-        this.setState({
-            view: "search"
-        });
-        return Promise.resolve();
     } else {
-       return Network.get("/rhn/manager/api/states/packages/match?sid=" + serverId + "&target=" + this.state.filter).promise.then(data => {
-          console.log(data);
-          this.setState({
-            view: "search",
-            search:  {
-                filter: this.state.filter,
-                results: data.map(state => {
-                  state.packageStateId = normalizePackageState(state.packageStateId);
-                  return state;
-                })
-            }
-          });
-        });
+      setChanged(draft => {
+        draft[key] = {
+          original: original,
+          value: {
+            arch: original.arch,
+            epoch: original.epoch,
+            version: original.version,
+            release: original.release,
+            name: original.name,
+            packageStateId: newPackageStateId,
+            versionConstraintId: newVersionConstraintId
+          }
+        };
+      });
     }
   }
 
-  save() {
-    const states = [];
-    for(var state of this.state.changed.values()) {
-        states.push(state.value)
+  const applyPackageState = () => {
+    if (changed.size > 0) {
+      const response = confirm(t("There are unsaved changes. Do you want to proceed ?"));
```

Comment:
The linter is complaining that we aren't using the confirm function in the right way!

Can't we just remove this part of the code? Since you disabled the button Apply when there are pending changes, I believe we don't need this anymore.

It could also be helpful to add a tooltip on hover on the button "apply" when there are pending change alerting the user to save them. This can be done simply by adding the attribute tooltip="message" when the button is disabled

Reply:
Yeah we can remove this part of the code. I will add the corresponding help for the buttons.

Reply:
Fixed in cc16218